### PR TITLE
[MIG] purchase_request: Migrate to version 17.0

### PR DESCRIPTION
**Pull request by sonhd91**

### CODE REVIEW

File: purchase_request/views/purchase_order_view.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='order_line']/tree" position="inside">
+                <field name="purchase_request_lines" invisible="1" />
+                <button
+                    attrs="{'invisible': [('purchase_request_lines', '=', [])]}"
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_order_view.xml]: remove `attrs`

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                <newline />
+                <group>
+                    <field name="item_ids" nolabel="1" colspan="2">
+                        <tree name="Details" create="false" editable="bottom">
+                            <field
+                                name="line_id"
+                                options="{'no_open': true}"
+                                invisible="1"
+                            />
+                            <field name="request_id" attrs="{'readonly': True}" />
```
Comment (nguyenminhchien) [inline purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml]: remove `attrs`

File: purchase_request/views/purchase_order_view.xml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+<!-- Copyright 2018-2019 ForgeFlow, S.L.
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='order_line']/tree" position="inside">
+                <field name="purchase_request_lines" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_order_view.xml]:
```suggestion
                <field name="purchase_request_lines" column_invisible="1" />
```

File: purchase_request/views/purchase_order_view.xml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            </field>
+        </field>
+    </record>
+    <record id="purchase_order_line_tree" model="ir.ui.view">
+        <field name="name">purchase.order.line.tree</field>
+        <field name="model">purchase.order.line</field>
+        <field name="inherit_id" ref="purchase.purchase_order_line_tree" />
+        <field name="arch" type="xml">
+            <xpath expr="//tree" position="inside">
+                <field name="purchase_request_lines" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_order_view.xml]:
```suggestion
                <field name="purchase_request_lines" column_invisible="1" />
```

File: purchase_request/views/purchase_order_view.xml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                <button
+                    invisible="purchase_request_lines == False"
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_order_view.xml]:
```suggestion
                    column_invisible="purchase_request_lines == False"
```

File: purchase_request/views/purchase_request_view.xml [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+                                    />
+                                    <field
+                                        name="analytic_distribution"
+                                        widget="analytic_distribution"
+                                        groups="analytic.group_analytic_accounting"
+                                        options="{'product_field': 'product_id', 'business_domain': 'purchase_order'}"
+                                    />
+                                    <field name="date_required" />
+                                    <field name="estimated_cost" widget="monetary" />
+                                    <field name="currency_id" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                                    <field name="currency_id" column_invisible="1" />
```

File: purchase_request/views/purchase_request_view.xml [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+                                    <field
+                                        name="company_id"
+                                        groups="base.group_multi_company"
+                                        widget="selection"
+                                    />
+                                    <field name="cancelled" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                                    <field name="cancelled" column_invisible="1" />
```

File: purchase_request/views/purchase_request_view.xml [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                                    <field name="is_editable" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                                    <field name="is_editable" column_invisible="1" />
```

File: purchase_request/views/purchase_request_view.xml [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+                                <tree
+                                    decoration-muted="cancelled == True"
+                                    editable="bottom"
+                                >
+                                    <field name="product_id" />
+                                    <field name="name" />
+                                    <field name="product_qty" />
+                                    <field
+                                        name="product_uom_category_id"
+                                        invisible="1"
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                                        column_invisible="1"
```

File: purchase_request/views/purchase_request_view.xml [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+                                    />
+                                    <field
+                                        name="product_uom_id"
+                                        invisible="1"
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                                        column_invisible="1"
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 11]
```diff
@@ -0,0 +1,590 @@
...
+                    widget="badge"
+                    decoration-success="request_state in ('done', 'approved')"
+                    decoration-info="request_state in ('draft', 'to_approve')"
+                    decoration-danger="request_state == 'rejected'"
+                />
+                <field name="requested_by" widget="many2one_avatar_user" />
+                <field name="assigned_to" widget="many2one_avatar_user" />
+                <field name="date_required" />
+                <field name="estimated_cost" widget="monetary" />
+                <field name="currency_id" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="currency_id" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 12]
```diff
@@ -0,0 +1,590 @@
...
+                    groups="analytic.group_analytic_accounting"
+                    options="{'product_field': 'product_id', 'business_domain': 'purchase_order'}"
+                />
+                <field name="supplier_id" />
+                <field
+                    name="company_id"
+                    groups="base.group_multi_company"
+                    widget="selection"
+                />
+                <field name="cancelled" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="cancelled" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 13]
```diff
@@ -0,0 +1,590 @@
...
+                <field name="qty_in_progress" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="qty_in_progress" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 14]
```diff
@@ -0,0 +1,590 @@
...
+                <field name="qty_done" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="qty_done" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 15]
```diff
@@ -0,0 +1,590 @@
...
+                <field name="qty_cancelled" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="qty_cancelled" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 16]
```diff
@@ -0,0 +1,590 @@
...
+                    decoration-success="purchase_state == 'done'"
+                    decoration-muted="purchase_state == 'draft'"
+                    decoration-info="purchase_state in ('sent', 'purchase')"
+                    decoration-warning="purchase_state == 'to_approve'"
+                    decoration-danger="purchase_state == 'cancelled'"
+                />
+                <field name="product_id" />
+                <field name="name" />
+                <field name="product_qty" />
+                <field name="product_uom_category_id" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="product_uom_category_id" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 17]
```diff
@@ -0,0 +1,590 @@
...
+                            <field name="purchase_request_allocation_ids">
+                                <tree decoration-muted="purchase_state == 'cancel'">
+                                    <field name="purchase_line_id" />
+                                    <field name="stock_move_id" />
+                                    <field name="product_id" />
+                                    <field name="requested_product_uom_qty" />
+                                    <field name="allocated_product_qty" />
+                                    <field name="open_product_qty" />
+                                    <field name="product_uom_id" />
+                                    <field name="purchase_state" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                                    <field name="purchase_state" column_invisible="1" />
```

File: purchase_request/views/purchase_request_line_view.xml [hunk 18]
```diff
@@ -0,0 +1,590 @@
...
+                <field name="order_id" />
+                <field name="name" />
+                <field name="partner_id" string="Vendor" />
+                <field name="product_id" />
+                <field name="price_unit" />
+                <field name="product_qty" />
+                <field name="product_uom" groups="uom.group_uom" />
+                <field name="price_subtotal" widget="monetary" />
+                <field name="date_planned" widget="date" />
+                <field name="purchase_request_lines" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_line_view.xml]:
```suggestion
                <field name="purchase_request_lines" column_invisible="1" />
```

File: purchase_request/views/purchase_request_view.xml [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+        <field name="name">purchase.request.tree</field>
+        <field name="model">purchase.request</field>
+        <field name="arch" type="xml">
+            <tree
+                decoration-bf="message_needaction==True"
+                decoration-info="state in ('draft','to_approve')"
+                decoration-muted="state in ('rejected')"
+                name="Purchase Request"
+            >
+                <field name="message_needaction" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                <field name="message_needaction" column_invisible="1" />
```

File: purchase_request/views/purchase_request_view.xml [hunk 20]
```diff
@@ -0,0 +1,403 @@
...
+                <field name="date_start" />
+                <field name="requested_by" widget="many2one_avatar_user" />
+                <field
+                    name="company_id"
+                    groups="base.group_multi_company"
+                    widget="selection"
+                />
+                <field name="activity_ids" widget="list_activity" optional="show" />
+                <field name="origin" />
+                <field name="currency_id" invisible="1" />
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_request_view.xml]:
```suggestion
                <field name="currency_id" column_invisible="1" />
```

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+                    <field name="supplier_id" />
+                </group>
+                <newline />
+                <group>
+                    <field name="item_ids" nolabel="1" colspan="2">
+                        <tree name="Details" create="false" editable="bottom">
+                            <field
+                                name="line_id"
+                                options="{'no_open': true}"
+                                invisible="1"
```
Comment (nguyenminhchien) [inline purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml]:
```suggestion
                                column_invisible="1"
```

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+                            />
+                            <field name="request_id" attrs="{'readonly': True}" />
+                            <field name="product_id" />
+                            <field name="name" />
+                            <field name="product_qty" />
+                            <field
+                                name="product_uom_id"
+                                invisible="1"
```
Comment (nguyenminhchien) [inline purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml]:
```suggestion
                                column_invisible="1"
```

File: purchase_request/__manifest__.py [hunk 23]
```diff
@@ -24,7 +24,7 @@
         "views/product_template.xml",
         "views/purchase_order_view.xml",
         "views/stock_move_views.xml",
-        "views/stock_picking_views.xml",
+        # "views/stock_picking_views.xml",
```
Comment (nguyenminhchien) [inline purchase_request/__manifest__.py]: remove this line and views/stock_picking_views.xml if you don't need them anymore

File: purchase_request/models/purchase_request.py [hunk 24]
```diff
@@ -240,15 +240,15 @@ def create(self, vals_list):
         for vals in vals_list:
             if vals.get("name", _("New")) == _("New"):
                 vals["name"] = self._get_default_name()
-        requests = super(PurchaseRequest, self).create(vals_list)
-        for vals, request in zip(vals_list, requests):
+        requests = super().create(vals_list)
+        for vals, request in zip(vals_list, requests, strict=False):
```
Comment (nguyenminhchien) [inline purchase_request/models/purchase_request.py]: should strict=True

File: purchase_request/views/stock_move_views.xml [hunk 25]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_stock_move_operations" model="ir.ui.view">
+        <field name="name">stock.move.operations.form</field>
+        <field name="model">stock.move</field>
+        <field name="inherit_id" ref="stock.view_stock_move_operations" />
+        <field name="arch" type="xml">
+            <field name="move_line_ids" position="after">
+                <newline />
+                <field name="purchase_request_allocation_ids" />
```
Comment (nguyenminhchien) [inline purchase_request/views/stock_move_views.xml:10]: ui broken
image

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]

**Comment by sonhd91** [comment]
Branch updated, please checked again

File: purchase_request/views/purchase_order_view.xml [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='order_line']/tree" position="inside">
+                <field name="purchase_request_lines" column_invisible="1" />
+                <button
+                    invisible="purchase_request_lines == False"
```
Comment (nguyenminhchien) [inline purchase_request/views/purchase_order_view.xml]:
```suggestion
                    column_invisible="purchase_request_lines == False"
```

**Review by nguyenminhchien — APPROVED** [review]
1 comment left, after updating you can create pr to oca

**Pull request merged**